Move About to a widget

edfloreshz opened the pull request:
Per @wiiznokes suggestion in #696, the About section code was moved to a widget.

Closes #696

mmstick commented:
Some tests are failing with an unresolved import for `about`

edfloreshz commented:
Sorry, should be fixed.

wiiznokes commented on src/widget/mod.rs:
```diff
@@ -365,3 +365,9 @@ pub use warning::*;
+pub mod about;
+#[cfg(feature = "desktop")]
+#[doc(inline)]
+pub use about::{about, About};
```
```suggestion
pub use about::about;
```

wiiznokes replied:
nevermind this, it's not a widget

edfloreshz replied:
I actually think this is a good idea.

wiiznokes commented on src/widget/about.rs:
```diff
@@ -117,3 +125,81 @@ impl About {
     }
 }
+
+#[cfg(feature = "desktop")]
```
this can be removed

edfloreshz replied:
Removed!

wiiznokes commented on src/widget/about.rs:
```diff
@@ -1,9 +1,17 @@
+    std::collections::BTreeMap,
+};
 
 #[cfg(feature = "desktop")]
```
this can be removed

edfloreshz replied:
Removed!

wiiznokes commented:
some nitpicks: 
- would be nice if the fields were private, and the `set_` prefix were removed.
- some data are duplicated. For example, repository_url is stored in a dedicated field and the links BtreeMap. I can see the value of having dedicated fields, for example, to add custom icons in the future
- would be nice to add a method `add_link(self, name, link) -> Self`
- `set_support_url` is ambigus imo, for example, i have this value set in metainfo.xml: `bugtracker`, `donation`, `vcs-browser`, `homepage`
- maybe switch `license` with `license_type`, and rename `license_type` to `licence_text`

wdyt @edfloreshz  ?

edfloreshz commented:
Those are great suggestions, I'll make the changes in the next couple of hours.

wiiznokes commented:
here are some ideas of urls we can add https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#url

edfloreshz commented:
- Properties are now private.
- Removed the URL specific methods and made method `links` public, letting anyone pass an array of label/link pairs.
- Added the `license` crate to handle licenses. 
- Removed `license_type` and replaced it with a link to the license page (https://spdx.org/licenses).

@wiiznokes

wiiznokes commented on src/widget/about.rs:
```diff
@@ -0,0 +1,195 @@
+    comments: Option<String>,
+    /// The application's copyright.
+    copyright: Option<String>,
+    /// The license text.
```
```suggestion
    /// The license name.
```

wiiznokes commented on src/widget/about.rs:
```diff
@@ -0,0 +1,195 @@
+    copyright: Option<String>,
+    /// The license text.
+    license: Option<String>,
+    /// Artists who contributed to the application.
```
Documentation for fields with `#[setters(skip)]` should be moved to the setter function instead.

wiiznokes commented on src/widget/about.rs:
```diff
@@ -0,0 +1,195 @@
+        let developers: BTreeMap<&'a str, &'a str> = developers.into();
+        self.developers = developers
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), format!("mailto:{v}")))
```
Could we provide a way to link something else than a email address ? Or just nothing. (maybe by checking if an `@` is present, or if the string is empty)

edfloreshz replied:
Those fields are meant for email addresses, if the user needs to provide a URL they can with `links`.

wiiznokes reviewed (approved):
I like the new api!
Just some minor doc issues

wiiznokes commented:
Could you add this method to support the `fl!` macro ?

```rs
pub fn link(mut self, name: impl Into<String>, link: &'a str) -> Self {
    self.links.insert(name.into(), link.to_string());
    self
}
```

edfloreshz commented:
I'm thinking of the best way to support localization with the existing set of methods.

wiiznokes commented:
> I'm thinking of the best way to support localization with the existing set of methods.

smtg like that could work:

```rs
pub fn links<const N: usize>(
            mut self,
            links: [(impl Into<String>, impl Into<String>); N],
        ) -> Self {
            self.links = links
                .into_iter()
                .map(|(k, v)| (k.into(), v.into()))
                .collect();
            self
        }
```

edfloreshz commented:
Decided to do this instead:

```rust
pub fn links<T: Into<String>>(mut self, links: impl Into<Vec<(T, &'a str)>>) -> Self {
    let links: Vec<(T, &'a str)> = links.into();
    self.links = links
        .into_iter()
        .map(|(k, v)| (k.into(), v.to_string()))
        .collect();
    self
}
```

wiiznokes commented:
Yeah, it works. The `Vec` also make probably more sense.

I noticed that the scrollbar is a bit off, but idk if it's related or not. Maybe we need to adjust the padding.

![Capture d’écran du 2024-11-09 21-08-58](https://github.com/user-attachments/assets/145bbb32-f6d6-4b3a-bd7d-1d098bae5f83)

edfloreshz commented:
Yeah we could give it some right padding

git-f0x commented:
Is there a reason that the scrollable was removed from the drawer? Since now it isn't possible to make the scrollbar be at the right edge (to match designs), and that spacing required to make the scrollbar not cover content makes things off-center.

wiiznokes commented:
> @wiiznokes Please confirm if I can ask Michael to review.

Yeah, go ahead

git-f0x commented:
If #698 gets merged, the scrollbar spacing can be removed, but `.padding([0, space_l, space_l, space_l])` should be applied to the column (before the scrollable).
